### PR TITLE
Functional Library with Example Demo

### PR DESCRIPTION
**Pull request by neildo**

### CODE REVIEW

File: job.go [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+	JobReader struct {
+		job     *Job
+		logs    io.ReadCloser
+		inotify *os.File
+	}
+)
+
+func Init() error {
+	// allow starting job
+	startable = true
```
Comment (nklaassen) [inline job.go]: messing with global variables is generally racy and hard to test, and why set it to true here if you might just set it back to false further down?

File: job.go [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+	if len(os.Args) < 3 && os.Args[1] == JailOp {
+		return ErrInvalidArgs
+	}
+	// skip run command in jail
+	if len(os.Args) < 3 || os.Args[1] != JailOp {
+		return nil
+	}
+	fmt.Println(os.Args)
```
Comment (nklaassen) [inline job.go]: why print, isn't this going to interfere with your job output?

File: job.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+	"io"
+	"os"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+)
+
+const (
+	JailOp = ".tjob"
```
Comment (nklaassen) [inline job.go]: nit: exported identifiers (constants, variables, types, functions, methods) should have GoDoc comments

File: job.go [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+	}
+	args := os.Args[2:]
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+	return fmt.Errorf(cmd.ProcessState.String())
```
Comment (nklaassen) [inline job.go]: why return this as an error?

File: job.go [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+// Stopped return true if StoppedAt is not zero
+func (s Status) Stopped() bool {
+	return !s.StoppedAt.IsZero()
+}
+
+// Start starts the command
+func (j *Job) Start(ctx context.Context) error {
+	if !startable {
+		panic("unsafe start")
```
Comment (nklaassen) [inline job.go]: panicking isn't very cool, could you return an error instead?

File: job.go [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+	}
+	// return any error from status if already stopped
+	if j.Status().Stopped() {
+		return j.status.Error
```
Comment (nklaassen) [inline job.go]: could you make the API prevent the possibility of starting the same job twice?

File: job.go [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+	}
+
+	// already started nothing to do
+	if j.Status().Started() {
+		return nil
```
Comment (nklaassen) [inline job.go]: isn't this kinda racy, couldn't two concurrent calls to job.Start() both get past this line?

File: job.go [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+		MemoryMB int
+
+		// ReadBPS represents the max bytes read per second by proc
+		ReadBPS int
+
+		// WriteBPS represents the max bytes write per second by proc
+		WriteBPS int
+
+		// Read-write lock for job state
+		rw sync.RWMutex
```
Comment (nklaassen) [inline job.go]: not very clear here which feels are/need to be protected by this mutex

File: job_linux.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+	// wait and ignore EOF until stopped
+	if n == 0 && err == io.EOF && !r.job.Status().Stopped() {
+
+		if n == 0 {
+			// watch for writes and close event
+			wd, err := syscall.InotifyAddWatch(int(r.inotify.Fd()), r.job.logs.Name(), syscall.IN_MODIFY|syscall.IN_CLOSE)
+			if err != nil {
+				return 0, os.NewSyscallError("inotify_add_watch", err)
+			}
+			defer syscall.InotifyRmWatch(int(r.inotify.Fd()), uint32(wd))
```
Comment (nklaassen) [inline job_linux.go]: why do you need to add and remove the watch on each read? isn't this a bit racy, you could read EOF, and then more could be written and the file closed before you add the watch, and then reading from the inotify file would block

File: job_linux.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+			// block until closed or write event
+			b := make([]byte, syscall.SizeofInotifyEvent*128)
+			if _, err = r.inotify.Read(b); err != nil {
+				return 0, err
```
Comment (nklaassen) [inline job_linux.go]: what could this error be and what does it mean to return it from your JobReader?

File: job_linux.go [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+			}
+		}
+		err = nil
```
Comment (nklaassen) [inline job_linux.go]: https://pkg.go.dev/io#Reader
> Implementations of Read are discouraged from returning a zero byte count with a nil error, except when len(p) == 0.

File: job_linux.go [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+	file := os.NewFile(uintptr(fd), l.Name())
+	go func() {
+		<-ctx.Done()
+		file.Close()
+	}()
+	return &JobReader{job: j, logs: l, inotify: file}, nil
+}
+
+// Read reads n bytes into buffer and return EOF only when Job stops
+func (r *JobReader) Read(buffer []byte) (n int, err error) {
```
Comment (nklaassen) [inline job_linux.go]: This is something that would be nice to have a unit test for

File: job.go (hunk 2 above)
Comment (neildo) [reply job.go]: good point to remove debug code

File: job.go (hunk 5 above)
Comment (neildo) [reply job.go]: good point

File: job_linux.go (hunk 12 above)
Comment (neildo) [reply job_linux.go]: Thank you for the suggestion. I'll add the unit test

File: job.go [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+
+// Started return true if StartedAt is not zero
+func (s Status) Started() bool {
+	return !s.StartedAt.IsZero()
+}
+
+// Stopped return true if StoppedAt is not zero
+func (s Status) Stopped() bool {
+	return !s.StoppedAt.IsZero()
+}
```
Comment (rosstimothy) [inline job.go:157]: What happens if either of these times are written to concurrently while consumers of the library call these functions?

File: job_darwin.go
Comment (rosstimothy) [inline job_darwin.go:1]: This might aid local testing, however, there is no requirement for this challenge that the system works on macOS. Consider removing it instead of having to support two separate implementations.

File: job_linux.go [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+	// create a directory structure like /sys/fs/cgroup/tjobs/<job_id>
+	if err := os.MkdirAll(cgroupJob, cgroupFileMode); err != nil {
+		return nil, fmt.Errorf("mkdir %s: %w", cgroupJob, err)
+	}
+
+	// enable cpu, io, and memory controllers
+	path := cgroupRoot + "/cgroup.subtree_control"
+	if err := os.WriteFile(path, []byte("+cpu +io +memory"), cgroupFileMode); err != nil {
+		return nil, fmt.Errorf("%s: %w", path, err)
```
Comment (rosstimothy) [inline job_linux.go:66]: If any of these branches return early nothing is being cleaned up from disk.

File: job.go (hunk 1 above)
Comment (neildo) [reply job.go]: You're right a bool is unclear on purpose. I renamed it to `procState` and use int32 to call `CompareAndSwapInt32` for advancing the state: notInited -> startable -> jailed. I want to prevent the undefined behavior of jailed procs from start a new job.

File: job.go (hunk 4 above)
Comment (neildo) [reply job.go]: I return error to force early exit by caller like this below.
```
func main() {
	if err := tjob.Init(); err != nil {
		fmt.Println(err.Error())
		return
	}
...
```

File: job_linux.go [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+		file.Close()
+	}()
+	return &JobReader{job: j, logs: l, inotify: file}, nil
+}
+
+// Read reads n bytes into buffer and return EOF only when Job stops
+func (r *JobReader) Read(buffer []byte) (n int, err error) {
+	n, err = r.logs.Read(buffer)
+	// wait and ignore EOF until stopped
+	if n == 0 && err == io.EOF && !r.job.Status().Stopped() {
```
Comment (rosstimothy) [inline job_linux.go]: Suggestion: handle the happy case and return early to unindent this code.

https://go.dev/wiki/CodeReviewComments#indent-error-flow

File: job_linux.go (hunk 9 above)
Comment (rosstimothy) [reply job_linux.go]: +1.

File: job.go (hunk 13 above)
Comment (neildo) [reply job.go:157]: `Job.Status()` locks the shared `status` before returning a copy of it. No contention with separate copies of Status.

File: job_darwin.go
Comment (neildo) [reply job_darwin.go:1]: done.

File: job.go (hunk 8 above)
Comment (neildo) [reply job.go]: I'll order the fields and clarify the comment

File: job_linux.go (hunk 10 above)
Comment (neildo) [reply job_linux.go]: You're right. I return only when closed by the context with `fs.ErrClosed`

File: job_linux.go (hunk 11 above)
Comment (neildo) [reply job_linux.go]: replace with `ErrReadAgain`

File: job.go [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+// Done returns true if the Job completed
+func (j *Job) Done() bool {
+	return j.state == stopped
+}
+
+// Logs returns JobReader for polling logs until process stops
+func (j *Job) Logs(ctx context.Context) (io.ReadCloser, error) {
+	if j.logs == nil {
+		return nil, ErrNotStartable
+	}
```
Comment (rosstimothy) [inline job.go]: What happens if two goroutines call Start and Logs on a single job instance at the same time?

File: job_linux.go [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+
+		// sufficiently size buffer for events
+		b := make([]byte, syscall.SizeofInotifyEvent*syscall.NAME_MAX+1)
+
+		// return EOF if file close by context
+		if n, err = r.inotify.Read(b); errors.Is(err, fs.ErrClosed) {
+			return 0, io.EOF
+		}
+		// keep reading
+		err = ErrReadAgain
```
Comment (rosstimothy) [inline job_linux.go]: So if a user of the library reads and receives an ErrReadAgain are they supposed to ignore the error and call Read again? Should they do so immediately? Should they wait first?

That contract seems a bit unidiomatic. Should this just try reading again itself?

File: job_test.go [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+func (m *JobMock) Done() bool {
+	return m.done
+}
+
+func TestJobReader(t *testing.T) {
+	tmp, err := os.CreateTemp("", "*")
+	if err != nil {
+		t.Fatalf("unexpected tmp file: %v", err)
+	}
+	defer func() { tmp.Close() }()
```
Comment (rosstimothy) [inline job_test.go]: Suggestion: use t.TempDir instead.

File: job_test.go [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+type JobMock struct {
+	tjob.Doner
+	done bool
+}
+
+func (m *JobMock) Done() bool {
+	return m.done
+}
+
+func TestJobReader(t *testing.T) {
```
Comment (rosstimothy) [inline job_test.go:26]: Could this test be extended, or another be added to cover other edge cases? The challenge spec specifically calls out having tests for both the happy and unhappy path.

File: job_linux.go (hunk 17 above)
Comment (neildo) [reply job_linux.go]: You're right. That would make the call easier to use and less error prone.
